File: src/main/java/com/kishlaly/ta/analyze/testing/tp/TakeProfitFixedKeltnerTop.java
package com.kishlaly.ta.analyze.testing.tp;

import com.kishlaly.ta.model.SymbolData;
import com.kishlaly.ta.model.indicators.KeltnerJava;
import com.kishlaly.ta.utils.IndicatorUtils;

/**
 * TP at % level from the middle to the top of the Keltner channel
 */
public class TakeProfitFixedKeltnerTop extends TakeProfitStrategy {

    public TakeProfitFixedKeltnerTop(Object config) {
        super(config, false);
    }

    @Override
    public double calcualte(SymbolData data, int signalIndex) {
        KeltnerJava keltner = IndicatorUtils.buildKeltnerChannels(data.symbol, data.quotes).get(signalIndex);
        int keltnerTopRatio = (int) getConfig();
        double middle = keltner.getMiddle();
        double top = keltner.getTop();
        double diff = top - middle;
        double ratio = diff / 100 * keltnerTopRatio;
        return middle + ratio;
    }

    @Override
    public String toString() {
        return "TP [Fixed] Keltner " + getConfig() + "% top ";
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/testing/tp/TakeProfitFixedKeltnerTop.kt
package com.kishlaly.ta.analyze.testing.tp

import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.utils.IndicatorUtils

/**
 * TP at % level from the middle to the top of the Keltner channel
 */
class TakeProfitFixedKeltnerTop(config: Any?) : TakeProfitStrategy(config) {

    override fun calculate(data: SymbolData, signalIndex: Int): Double {
        val keltner = IndicatorUtils.buildKeltnerChannels(data.symbol, data.quotes)[signalIndex]
        val keltnerTopRatio = config as Double
        val middle = keltner.middle
        val top = keltner.top
        val diff = top - middle
        val ratio = diff / 100 * keltnerTopRatio
        return middle + ratio
    }

    override fun toString() = "TP [Fixed] Keltner ${config as Double}% top"

}